lib/fast_jsonapi/serialization_core.rb

```diff
@@ -68,23 +67,38 @@ def meta_hash(record, params = {})
       def record_hash(record, fieldset, includes_list, params = {})
         if cache_store_instance
           cache_opts = record_cache_options(cache_store_options, fieldset, includes_list, params)
-          record_hash = cache_store_instance.fetch(record, **cache_opts) do
-            temp_hash = id_hash(id_from_record(record, params), record_type, true)
-            temp_hash[:attributes] = attributes_hash(record, fieldset, params) if attributes_to_serialize.present?
-            temp_hash[:relationships] = relationships_hash(record, cachable_relationships_to_serialize, fieldset, includes_list, params) if cachable_relationships_to_serialize.present?
-            temp_hash[:links] = links_hash(record, params) if data_links.present?
-            temp_hash
+          klass = self.name
+          param = [record, cache_opts, klass]
+          ::BatchLoader.for(param).batch do |batch_params, loader|
+            cache_results = cache_store_instance.fetch(batch_params)
+              .map { |str| JSON.parse(str, symbolize_names: true) }.group_by { |h| [h[:id], h[:type]] }.transform_values(&:first)
+            batch_params.each do |batch_param|
+              _record = batch_param.first
+              _cache_opts = batch_param.second
+              _klass = Object.const_get(batch_param.last)
+              record_h = cache_results[[_record.id.to_s, _record.class.model_name.name.underscore]]
```

Comment:
`mget` preserves sequence in returned value based on input so you don't need to build a map. just use the index to match the output with input should work